### DISCUSSION TITLE
Questions about Lite Self hosted version

### DISCUSSION THREAD

**Discussion by aejazm** [gorules/zen#97] [General]
Hi,

I want to try out the rule engine & would like more information before proceeding.
- How are the rules loaded dynamically ?
- How is the horizontal scalability achieved ?
- Is it production ready ?

Thanks

**Comment by stefan-gorules** [comment]
Hi @aejazm, thank you for your interest. I'll provide answers here, but keep in mind that Zen is our open-source project - core of rules engine, whereas BRMS (what you are inquiring about) is commercial. 

Regarding your questions:
1. Could you clarify what you are referring to in a bit more detail? If you mean how they are loaded, it's directly from the database. Broadly speaking, every evaluation = 1 SQL Select + execution.
2. Our BRMS can scale both vertically and horizontally. To scale horizontally the easiest way is to can use Kubernetes, Azure Container Apps or similar technologies. DB on the other hand can be scaled with read replicas.
3. Self-hosted plans are production-ready. You need to ensure your infrastructure meets your scaling needs.

**Reply by aejazm** [reply]
Hi Stefan,

Thanks for your reply. Sorry that I wasn't clear before. Here is little bit more detail about my questions which were mostly regarding the core zen engine. As I understand, once the engine instance is up with a graph json containing the decision table(s), if I want to add one or more rows to the decision table, do I need to restart the engine with the updated graph JSON or I can reload the new rows without stopping the engine.

**Reply by stefan-gorules** [reply]
Regarding core engine - it depends on your approach. Here are few examples from node.js:

```ts
import { ZenEngine } from '@gorules/zen-engine';

const engine = new ZenEngine();

// as long as "loadFromDatabase" gives new updated graphJson, no need to restart service
const reloadEveryTime = async (context: unknown) => {
   const graphJson = await loadFromDatabase();
   const decision = engine.createDecision(graphJson);
   
   return decision.evaluate(context);
}

// For example, this would never reload until service is restarted
const cachedGraphJson = Buffer.from(...);
const cachedDecision = engine.createDecision(cachedGraphJson);
const alwaysCached = async (context: unknown) => {
  return cachedDecision.evaluate(context);
}
```

On other hand, in case of BRMS, all you have to do is make sure that new version is published, and not just updated.

**Reply by ivanmiletic** [reply]
Hi @aejazm, in short, you don't need to restart table, one of the core benefits of zen-engine is loading models, parsing and compiling on the fly in sub millisecond time. You just need to update your source file and with the next execution it will work.

For complete BRMS, it does it automatically for you.

**Reply by aejazm** [reply]
Thanks Stefan & Ivan. That does clarifies things. I will give it a try.
